### ISSUE TITLE
bug report: GetILFunctionBodyAllocator() wrapper makes SetRVA() call meaningless

### ISSUE THREAD

**Issue by sanikolov** [microsoft/CLRInstrumentationEngine#192]
Scenario: define your own method, set the IL and set the RVA.
Observed behavior: there is no user visible facility by which the RVA can be set because wrapper GetILFunctionBodyAllocator() returns a bogus allocator and body address. There is no alternative in place to set the RVA.
Desired behavior: make it possible for API users to emit their own methods and set the RVA correctly.
Priority: high

```
CComPtr<IMetaDataEmit2> pMetaDataEmit;
pModuleInfo->GetMetaDataEmit(reinterpret_cast<IUnknown**>(&pMetaDataEmit));
....
pMetaDataEmit->DefineMethod(....a bunch of params here...., &methodToken);
CComPtr<ICorProfilerInfo2> profilerInfo;
pProfilerManager->GetCorProfilerInfo(reinterpret_cast<IUnknown**>(&profilerInfo));
profilerInfo->GetILFunctionBodyAllocator(moduleId, &pMethodAlloc); // bogus allocator returned
const BYTE *newBody = ExportIL(pMethodAlloc, etc, ...);
ULONG rva = reinterpret_cast<uintptr_t>(newBody) & ~0UL;
pMetaDataEmit->SetRVA(methodToken, rva); // rva param is garbage
profilerInfo->SetILFunctionBody(moduleId, methodToken, newBody);
```

See also
- [CorProfilerInfoWrapper.cpp#L387](https://github.com/microsoft/CLRInstrumentationEngine/blob/3c15dd95474c260e047080a9ef0bf446cf3dd377/src/InstrumentationEngine/CorProfilerInfoWrapper.cpp#L387)
- [CorMethodMalloc.h#L10](https://github.com/microsoft/CLRInstrumentationEngine/blob/cac2e272e1a5f4d8046c653f9f418a4a90d853d4/src/InstrumentationEngine/CorMethodMalloc.h#L10)

**Comment by delmyers** [comment]
@sanikolov Out of curiosity, why is it that you are attempting to set the RVA in the first place? 

In general, when you are using the ClrInstrumentationEngine, most of your IL trasformations should not use the profiler info directly, unless your instrumentation method has been set up as a "Raw Profiler", of which there can only be one. We supply a wrapper to the ICorProfilerInfo for some low-level querying about profiler information, but for the most part, you shouldn't be calling these. Also, the RVA for the defined method will be set by the runtime at Jit time. You don't really have a choice in the matter.

There is a known bug in the CLR that does not allow you to define methods with invalid RVAs that will get set at Jit time, though. We normally work around this in our own code using the following process:

1) set methodRva = some pre-existing rva from any method in the assembly. It doesn't matter which, the CLR will just reset the rva any way.
2) call pMetaDataEmit->DefineMethod(..., methodRva, implFlags, &token)
3) emit the function body at Jit time, using ClrInstrumentationEngine apis.
4) The Clr will internally update the RVA.

Potentially, what we should do here is add a public API to IModuleInfo, or ITypeInfo so that users of ClrInstrumentationEngine don't need to do this kind of workaround?

@wiktork @WilliamXieMSFT @jander-msft , what do you think?

**Comment by sanikolov** [comment]
I tried leaving out the RVA. It was specified as 0 in DefineMethod.
That resulted in
```
Unhandled Exception: System.TypeLoadException: Could not load type 'BlahType' from assembly 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' because the method 'Blah' has no implementation (no RVA).
```
I will try your suggestion and if it works then this ticket can be closed.

**Comment by sanikolov** [comment]
@delmeyers Can you elaborate what you mean by **3. emit the function body at Jit time, using ClrInstrumentationEngine apis.** ?
Suppose I DefineMethod() "BlahMethod" and get a token for it.
I am not aware of any CLRIE API that will result in a callback to {Before,Should,""}InstrumentMethod.
So how am I supposed to specify the IL for "BlahMethod" with IInstruction, or some other interface?

**Comment by wiktork** [comment]
@sanikolov The flow should be something like this:
1. During ModuleLoad you call DefineMethod and save off the created token.
2. During Rejit/Jit for an existing method, you will typically instrument a Call instruction against the token you created in step 1.
3. Once a method you are instrumenting calls the method you created in step 1, the Clrie will give you a ShouldInstrumentMethod callback (at jit time). You can check the token against the token you created in step 1, and indicate that the method needs instrumentation.
4. You will now get a callback to InstrumentMethod for your new method. You can fill out the instruction graph with the appropriate instructions. 

As you can see, there's no explicit InstrumentMethod callback invocation; it's done automatically at Jit Time. This means you must instrument something that calls your new method in existing code. Forcing a rejit on an existing method can also be helpful here.

You can use the IDataContainer::GetDataItem and IDataContainer::SetDataItem on your IModuleInfo object to make stashing module level information easier.

**Comment by sanikolov** [comment]
@wiktork many thanks. I know how to do what you are suggesting. Let's keep this ticket open until I pull this off and then we can close it.

**Comment by sanikolov** [comment]
I followed the recommendation in this ticket and have a couple observations to add.

If the RVA in DefineMethod() is set to 0 then I get
Unhandled Exception: System.TypeLoadException: Could not load type 'BlahType' from assembly 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' because the method 'Blah' has no implementation (no RVA).
**Please note that mscorlib is not special in this case**. The issue applies equally to any module in which a DefineMethod() occurs.

If the RVA in DefineMethod() is set to some arbitrary valid RVA (within the module), as recommended earlier in this ticket,
then in callback InstrumentMethod() I attempt to fill out the instruction graph but that fails. When I print the instruction graph
*without* instrumenting the method, the printout shows the instructions of the method pointed to by the RVA I specified in DefineMethod().
When I attempt to change the IL then I get
`IfFailRet(igraph->InsertBeforeAndRetargetOffsets(current, instr)) failure with hresult 0x80004002`.
The instructions are inserted backward from the beginning, i.e. [new_first, 2nd, 3rd.... originalFirst, originalSecond....last] .
This method of insertion has never failed before for other IL bodies that I've changed.
0x80004002 maps to E_NOINTERFACE.
In other words, despite seeing the newly defined method appear in ShouldInstrumentMethod() and InstrumentMethod() as a result of calling it, something is different
about it. The standard technique of filling out the instruction graph, which I've used successfully in many other cases, doesn't apply.

**Comment by delmyers** [comment]
@sanikolov, It is expected that you cannot set the RVA to 0 in DefineMethod(). As stated earlier "There is a known bug in the CLR that does not allow you to define methods with invalid RVAs"

I'm a little confused by what you are saying here:

> IfFailRet(igraph->InsertBeforeAndRetargetOffsets(current, instr)) failure with hresult 0x80004002.
The instructions are inserted backward from the beginning, i.e. [new_first, 2nd, 3rd.... originalFirst, originalSecond....last] .

If you are inserting instructions, each above the last, it seems that I would expect that they would all end up in reversed order. If you are trying to insert instructions from top to bottom, why aren't you using InsertAfter? Or are you trying to insert them backwards on purpose?

Also, calling InsertBeforeAndRetargetOffsets(...) for an empty function will put everything in the function into the prolog (the bit of code that is meant to set up parameters, etc, before your actual function executes). I don't know if the CLR will have problems with that or not. I don't think it will make a difference, but it is unusual.

You could be getting E_NOINTERFACE for a number of reasons. For example, if you create an OpCode  instruction, and give it a branch opcode, but don't create the branch instruction using CreateBranchInstruction, then when the algorithm attempts to retarget branch instructions, it will fail to cast your instruction to the correct type and return E_NOINTERFACE. A similar thing could happen with switch statements. Do you have a minimum bit of code that will repro the issue?

Note, if you are creating branch instructions, you are allowed to pass NULL to the second parameter. That is, you don't have to know the branch target up-front. The documentation and annotation on that interface is incorrect. I'll file a bug.

**Comment by sanikolov** [comment]
Never mind, instructionGraph->RemoveAll()
is required as I found out this morning. Without it I get 0x8004002 and stale instructions.
No insertion errors if I RemoveAll() first. Odd.
Thanks for all the help. Closing ticket.

**Comment by delmyers** [comment]
Added bug #193 to update the documentation of CreateBranchInstruction
